File: week6/sentimental-credit/credit.py
while True:
    try:
        number = int(input("Number: "))
        break
    except:
        continue

str_number = str(number)
length = len(str_number)

# Luhn’s Algorithm
# Step 1 : Multiply every other digit by 2, starting with the number’s second-to-last digit,
# and then add those products’ digits together.
SumEveryTwox2 = 0
for i in range(1, length, 2):
    EveryTwox2 = int(str_number[length-i-1])*2

    if EveryTwox2 < 10:
        SumEveryTwox2 = SumEveryTwox2 + EveryTwox2
    else:
        str_tens = str(EveryTwox2)
        SmallSum = 1 + int(str_tens[1])
        SumEveryTwox2 = SumEveryTwox2 + SmallSum

# Step 2 : Add the sum to the sum of the digits that weren’t multiplied by 2.
SumEveryOtherTwo = 0
for j in range(0, length, 2):
    EveryOtherTwo = int(str_number[length-j-1])
    SumEveryOtherTwo = SumEveryOtherTwo + EveryOtherTwo

# Step 3 : Add the two sums from the Luhn’s Algorithm
verify_number = SumEveryTwox2 + SumEveryOtherTwo

# Step 4 : Condition if card number is valid, AMEX, Mastercard or Visa
if ((verify_number % 10) != 0):
    print("INVALID")
    exit(0)

# First & first2 digit
first = int(number / pow(10, length - 1))
first2 = int(number / pow(10, length - 2))

if length == 15 and (first2 == 34 or first2 == 37):
    print("AMEX")
elif length == 16 and (first2 == 51 or first2 == 52 or first2 == 53 or first2 == 54 or first2 == 55):
    print("MASTERCARD")
elif (length == 13 or length == 16) and first == 4:
    print("VISA")
else:
    print("INVALID")

File: week1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int WhatDigit(long number, int n);

int main(void)
{
    // Credit card number
    long cardNumber = get_long("What is the card number you want to check ?: ");
    printf("Number : %li\n", cardNumber);

    //Total number of digit
    double cardLength = trunc(log10(cardNumber) + 1);

    //First digit
    int first = WhatDigit(cardNumber, cardLength - 1);

    //First 2 digit
    int first2 = cardNumber / pow(10, cardLength - 2);

    // Luhn’s Algorithm
    // Every two digit from the second last digit
    int EveryTwox2 = 0;
    int SumEveryTwox2 = 0;
    long cardLengthTrunc = trunc(cardLength / 2);
    for (int i = 0, x = 1; i < cardLengthTrunc; ++i, x = x + 2)
    {
        EveryTwox2 = (WhatDigit(cardNumber, x)) * 2;

        if (EveryTwox2 < 10)
        {
            SumEveryTwox2 = SumEveryTwox2 + EveryTwox2;
        }
        else
        {
            int UnitsDigit = WhatDigit(EveryTwox2, 0);
            int SmallSum = 1 + UnitsDigit;

            SumEveryTwox2 = SumEveryTwox2 + SmallSum;
        }
    }

    // Every two digit from the last digit
    int EveryOtherTwo = 0;
    int SumEveryOtherTwo = 0;
    long cardLengthCeil = ceil(cardLength / 2);
    for (int j = 0, y = 0; j < cardLengthCeil; ++j, y = y + 2)
    {
        EveryOtherTwo = WhatDigit(cardNumber, y);
        SumEveryOtherTwo = SumEveryOtherTwo + EveryOtherTwo;
    }

    // Add the two sums from the Luhn’s Algorithm
    int VerifyNumber = SumEveryTwox2 + SumEveryOtherTwo;

    // Condition if card number is valid, AMEX, Mastercard or Visa
    if ((VerifyNumber % 10) != 0)
    {
        printf("INVALID\n");
    }
    else
    {
        // Conditions : AMEX or Mastercard or Visa
        if (cardLength == 15 && (first2 == 34 || first2 == 37))
        {
            printf("AMEX\n");
        }
        else if (cardLength == 16 && (first2 == 51 || first2 == 52 || first2 == 53 || first2 == 54 || first2 == 55))
        {
            printf("MASTERCARD\n");
        }
        else if ((cardLength == 13 || cardLength == 16) && first == 4)
        {
            printf("VISA\n");
        }
        else
        {
            printf("INVALID\n");
        }
    }
}

//Function : Get a specific digit from an number (0 = last number, 0+1 = second last, etc)
int WhatDigit(long number, int n)
{
    long removeN = trunc(number / pow(10, n));
    int selectDigit = removeN % (10);
    return selectDigit;
}